doltpy/sql/sync/dolt.py

```diff
@@ -0,0 +1,237 @@
+    """
+    Returns a function that takes a list of tables and returns a mapping from the table name to the data returned by
+    the passed reader. The reader is generally one of `get_table_reader_diffs` or `get_table_reader`, but it would
+    be easy enough to provide some other kind of function if neither of these meet your needs.
+    :param repo:
+    :param reader:
+    :return:
+    """
+    def inner(tables: List[str]) -> DoltAsSourceUpdate:
+        result = {}
+        repo_tables = [table.name for table in repo.ls()]
+        missing_tables = [table for table in tables if table not in repo_tables]
+        if missing_tables:
+            logger.error('The following tables are missign, exiting:\n{}'.format(missing_tables))
+            raise ValueError('Missing tables {}'.format(missing_tables))
+
+        for table in tables:
+            logger.info('Reading table: {}'.format(table))
+            result[table] = reader(table, repo)
+
+        return result
+
+    return inner
+
+
+def get_table_reader_diffs(commit_ref: str = None, branch: str = None) -> Callable[[str, Dolt], DoltTableUpdate]:
+    """
+    Returns a function that reads the diff from a commit and/or branch, defaults to the HEAD of the current branch if
+    neither are provided.
+    :param commit_ref:
+    :param branch:
+    :return:
+    """
+    def inner(table_name: str, repo: Dolt) -> DoltTableUpdate:
+        current_branch, _ = repo.branch()
+        if branch and branch != current_branch:
+            repo.checkout(branch)
+
+        from_commit, to_commit = get_from_commit_to_commit(repo, commit_ref)
+        metadata = MetaData(bind=repo.get_engine())
+        metadata.reflect()
+        table = metadata.tables[table_name]
+        pks_to_drop = get_dropped_pks(repo.get_engine(), table, from_commit, to_commit)
+        result = _read_from_dolt_diff(repo.get_engine(), table, from_commit, to_commit)
+        return pks_to_drop, result
+
+    return inner
+
+
+def get_dropped_pks(engine: Engine, table: Table, from_commit: str, to_commit: str) -> List[dict]:
+    """
+    Given table_metadata, a connection, and a pair of commits, will return the list of pks that were dropped between
+    the two commits.
+    :param engine:
+    :param table:
+    :param from_commit:
+    :param to_commit:
+    :return:
+    """
+    pks = [col.name for col in table.columns if col.primary_key]
+    query = '''
+        SELECT
+            {pks}
+        FROM
+            dolt_diff_{table_name}
+        WHERE
+            from_commit = '{from_commit}'
+            AND to_commit = '{to_commit}'
+            AND diff_type = 'removed'
+    '''.format(pks=','.join(['`from_{}` as {}'.format(pk, pk) for pk in pks]),
+               table_name=table.name,
+               from_commit=from_commit,
+               to_commit=to_commit)
+
+    return _query_helper(engine, query)
+
+
+def get_from_commit_to_commit(repo: Dolt, commit_ref: str = None) -> Tuple[str, str]:
```

Comment:
some of these seem extremely specific to me, what's the use case for this?

Reply:
is it diffing?

Reply:
When syncing from Dolt to another database we want to grab only the data between two commits, that is every sync call grabs only the data changes since the last commit.

Reply:
👍 yeah that makes perfect sense